runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java

```diff
@@ -131,91 +131,105 @@ public void gc(long trimMark) {
+                        : runtime.getSequencerView().next(conflictInfo, streamIDs); // Token w/ conflict info
+
+                // Is our token a valid type?
+                AbortCause abortCause = null;
+                switch (tokenResponse.getRespType()) {
+                    case TX_ABORT_CONFLICT:
+                        abortCause = AbortCause.CONFLICT;
+                        break;
+                    case TX_ABORT_NEWSEQ:
+                        abortCause = AbortCause.NEW_SEQUENCER;
+                        break;
+                    case TX_ABORT_SEQ_OVERFLOW:
+                        abortCause = AbortCause.SEQUENCER_OVERFLOW;
+                        break;
+                    case TX_ABORT_SEQ_TRIM:
+                        abortCause = AbortCause.SEQUENCER_TRIM;
+                        break;
                 }
 
-                // Attempt to write to the log.
-                runtime.getAddressSpaceView().write(tokenResponse, ld, cacheOption);
-                // If we're here, we succeeded, return the acquired token.
-                return tokenResponse.getSequence();
-            } catch (OverwriteException oe) {
-                // We were overwritten, get a new token and try again.
-                log.warn("append[{}]: Overwritten after {} retries, streams {}",
-                        tokenResponse.getSequence(), x,
-                        Arrays.stream(streamIDs).map(Utils::toReadableId).collect(Collectors.toSet()));
-
-                if (conflictInfo != null) {
-                    // On retry, check for conflicts only from the previous attempt position,
-                    // otherwise the transaction will always conflict with itself.
-                    conflictInfo.setSnapshotTimestamp(tokenResponse.getToken());
+                if (abortCause != null) {
+                    throw new TransactionAbortedException(
+                            conflictInfo,
+                            tokenResponse.getConflictKey(), tokenResponse.getConflictStream(),
+                            tokenResponse.getToken().getSequence(), abortCause,
+                            TransactionalContext.getCurrentContext());
                 }
 
-            } catch (StaleTokenException se) {
-                // the epoch changed from when we grabbed the token from sequencer
-                log.warn("append[{}]: StaleToken, streams {}", tokenResponse.getSequence(),
-                        Arrays.stream(streamIDs).map(Utils::toReadableId).collect(Collectors.toSet()));
-
-                throw new TransactionAbortedException(
-                        conflictInfo,
-                        tokenResponse.getConflictKey(), tokenResponse.getConflictStream(),
-                        tokenResponse.getToken().getSequence(),
-                        AbortCause.NEW_SEQUENCER, // in the future perhaps define a new AbortCause?
-                        TransactionalContext.getCurrentContext());
+                try {
```

Comment:
everything inside for cycle above this point should be extracted into a method, let's say checkAbort();
```
for (int retry = 0; retry < runtime.getParameters().getWriteRetry(); retry++) {
                checkAbort();

                try {
                    if (TransactionalContext.isInTransaction()) {...
```

Reply:
the whole for cycle would be good to have like this:
```
for (int retry = 0; retry < runtime.getParameters().getWriteRetry(); retry++) {
                checkAbort();
                write();
```

Reply:
Done.